File: common/src/main/java/me/xanium/gemseconomy/GemsEconomy.java
package me.xanium.gemseconomy;

import me.xanium.gemseconomy.account.AccountManager;
import me.xanium.gemseconomy.currency.CurrencyManager;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

@SuppressWarnings("unused")
public class GemsEconomy {
    private static final GemsEconomy INSTANCE = new GemsEconomy();

    private CurrencyManager currencyManager;
    private AccountManager accountManager;

    public static @NotNull GemsEconomy getInstance() {
        return Objects.requireNonNull(INSTANCE, "api not initialized yet");
    }

    public AccountManager getAccountManager() {
        return accountManager;
    }

    public CurrencyManager getCurrencyManager() {
        return currencyManager;
    }

    @ApiStatus.Internal
    public void setAccountManager(@NotNull AccountManager accountManager) {
        Objects.requireNonNull(accountManager, "accountManager");
        this.accountManager = accountManager;
    }

    @ApiStatus.Internal
    public void setCurrencyManager(@NotNull CurrencyManager currencyManager) {
        Objects.requireNonNull(currencyManager, "currencyManager");
        this.currencyManager = currencyManager;
    }

    @ApiStatus.Internal
    public void unregister() {
        accountManager = null;
        currencyManager = null;
    }
}

File: common/src/main/kotlin/me/xanium/gemseconomy/GemsEconomy.kt
package me.xanium.gemseconomy

import org.jetbrains.annotations.ApiStatus
import java.util.*

interface GemsEconomy {
    /**
     * Get an account, or create one if not existing.
     *
     * @param uuid account's unique ID
     * @return an account object with specific unique ID
     */
    @ApiStatus.Experimental
    fun pullAccount(uuid: UUID): Account

    /**
     * Get an account, or return null if not existing.
     *
     * @param uuid account's unique ID
     * @return an account object with specific unique ID
     */
    @ApiStatus.Experimental
    fun getAccount(uuid: UUID): Account?

    /**
     * Check if specific account exists in database.
     *
     * @param uuid account's unique ID
     * @return true if account with specific unique ID exists in database
     */
    fun hasAccount(uuid: UUID): Boolean

    /**
     * Deposit specified amount into specific account.
     *
     * @param uuid   the account's unique ID
     * @param amount the amount of default currency
     */
    fun deposit(uuid: UUID, amount: Double)

    /**
     * Deposit specified amount into specific account.
     *
     * @param uuid     the account's unique ID
     * @param amount   the amount of specific currency
     * @param currency the specific currency
     */
    fun deposit(uuid: UUID, amount: Double, currency: Currency)

    /**
     * Withdraw specific amount from specific account.
     *
     * @param uuid   the account's unique ID
     * @param amount the amount of default currency
     */
    fun withdraw(uuid: UUID, amount: Double)

    /**
     * Withdraw specific amount from specific account.
     *
     * @param uuid     the account's unique ID
     * @param amount   the amount of specific currency
     * @param currency the currency you withdraw from
     */
    fun withdraw(uuid: UUID, amount: Double, currency: Currency)

    /**
     * Lookup the balance of specific account.
     *
     * @param uuid the account's unique ID
     * @return the balance of default currency
     */
    fun getBalance(uuid: UUID): Double

    /**
     * Lookup the balance of specific account.
     *
     * @param uuid     the account's unique ID
     * @param currency the amount of default currency
     * @return the balance of specific currency
     */
    fun getBalance(uuid: UUID, currency: Currency): Double

    /**
     * @param name the currency name
     * @return a currency object of specific name
     */
    @ApiStatus.Experimental
    fun getCurrency(name: String): Currency?

    /**
     * @return the default currency
     */
    @ApiStatus.Experimental
    fun getDefaultCurrency(): Currency

    /**
     * @return an unmodifiable list of loaded currencies
     */
    @ApiStatus.Experimental
    fun getLoadedCurrencies(): List<Currency>
}